rfd/0xxx-tls-routing-behind-lb.md

```diff
@@ -0,0 +1,319 @@
+  │ local credentials     │                            │
+  ├──────────────────────►│  server kube-xxx.xxx.xxx   │
+  │                       │  "real" credentials        │
+  │                       ├───────────────────────────►│
+  │                       │                            │
+  │                       │◄───────────────────────────┤
+  │◄──────────────────────┤                            │
+  │                       │                            │
+  │ another request       │                            │
+  ├──────────────────────►│                            |
+  │                       ├───────────────────────────►│
+  ...                   
+```
+
+Once the local server receives a request and verifies the request, the local server will proxy the request using the TLS
+Routing dialer to connect the Teleport Proxy. The local proxy should be in charge of managing the Teleport cert used for
+routing Kubernetes requests.
+
+Note that `kube-teleport-proxy-alpn.my-teleport-cluster.com` is not required to be resolvable (by DNS) in this case, but
+it will be used internally for routing.
+
+### User Experience
+
+Once the connection upgrade support to all protocols are implemented, users can be recommended to upgrade their Teleport
+clusters to single port mode if the cluster is currently serving separate ports.
+
+All scenarios discussed in this section assumes Teleport sits behind a load balancer so connection upgrade for TLS
+Routing is required.
+
+#### 1 - Database Access UX
+
+Common flows already use `tsh` local proxy (since v8) so they will automatically perform the connection upgrade:
+
+- `tsh db connect` (single port mode)
+- `tsh proxy db`
+
+So no UX change to these.
+
+In single port mode, using native clients to connect to Teleport Proxy directly with `tsh db env/config` will NOT work
+through load balancers. Users can use one of the above methods instead.
+
+#### 2 - Application Access UX
+
+Application Access through Teleport Webapp is NOT affected by the issue of interest of this RFD.
+
+On the `tsh` side, TCP, AWS, Azure, and GCP apps always use a local proxy so they will automatically perform the
+connection upgrade when needed. No UX change to these.
+
+However, HTTP apps MUST use `tsh proxy app`.
+
+#### 3 - Kubernetes Access UX
+
+When upgrade is required, using native Kubernetes clients after `tsh kube login` will NOT work as the native clients
+cannot upgrade the connection. `tsh kube login` should advertise users to run `tsh kubectl` and `tsh proxy kube` instead
+of `kubectl version`. `tsh kube credentials` should also error out and suggest using `tsh kubectl` and `tsh proxy kube`.
+
+`tsh kubectl` will automatically starts a local proxy and performs the connection upgrade when needed.
+
+A new `tsh proxy kube` command will be added to support other Kubernetes clients. The command will start a local proxy
+and provide a config for Kubernetes clients to use:
+
+```
+$ tsh proxy kube cluster1 cluster2 -p 8888
```

Comment:
Will we force users to identify the clusters they want to access when they start the proxy or will we allow them to select some (and we automatically generate the certs) and if a new cluster is requested, the proxy can create the new certificate on the fly and do the MFA if required?

Reply:
Good question. I initially thought we can apply [DB MFA session](https://github.com/gravitational/teleport/blob/master/rfd/0090-db-mfa-sessions.md) solution to the local kube proxy (where it prompt MFA once when starting local proxy). But I neglected the fact MFA is required per kube cluster.

One option is, when per-session-MFA is enabled, `tsh kube credentials` will ask the corresponding local proxy if they have cached MFA cert for the requested kube cluster. If not, `tsh kube credentials` will prompt for MFA and forward it to the local proxy. The prompt should happen only once per kube cluster per local proxy session.

What do you think?

I am curious how well the MFA prompt from `tsh kube credentials` works today for various kube applications.